### ISSUE TITLE
How to run the Node built-in testrunner for TypeScript files inside a specific directory?

### ISSUE THREAD

**Issue by jtuchel** [nodejs/help#3902]
### Details

I want to replace Mocha tests with the built-in testrunner. All tests are inside a test directory and follow the pattern

> ...Tests.ts

I started with a fooTests.ts file

```ts
import assert from 'assert/strict';
import test from 'node:test';

test('1 is equal to 1.', () => {
  assert.strictEqual(1, 1);
});
```

and added the npm script

`"test": "node --test ./test/**/*Tests.ts",`

The script fails with the message

> Could not find '/home/.../test/**/*Tests.ts'

How can I fix the script to

- test files matching the path `./test/**/*Tests.ts` only
- run with TS files ( I'm pretty sure I must use ts-node for this )

? Thanks in advance!

### Node.js version

v18

### Operating system

Linux ( Ubuntu 22.04 )

### Scope

Built-in testrunner

### Module and version

Not applicable.

**Comment by neodon** [comment]
In case it helps anyone who runs across this issue, here is how I currently execute tests written in TypeScript with the new node test runner and [esbuild-kit/tsx](https://github.com/esbuild-kit/tsx):

```bash
npm install -D tsx
```

Add the following npm script entry:

```json
"test": "node --loader tsx --test test/**/*Test.ts"
```

Let me know if this is helpful.

**Comment by jtuchel** [comment]
@neodon I tried to reproduce your approach but I get a "not found" error ( I'm using Node 19 )

**tsconfig.json**

```json
{
    "compilerOptions": {
        "baseUrl": ".",
        "declaration": true,
        "esModuleInterop": true,
        "lib": ["es2022", "dom"],
        "module": "commonjs",
        "outDir": "build",
        "resolveJsonModule": true,
        "strict": true,
        "target": "es2022"
    },
    "include": ["./**/*.ts"],
    "exclude": ["./build"]
}
```

**package.json**

```json
{
    "scripts": {
        "test": "tsc --noEmit && node --loader tsx --test test/**/*Tests.ts"
    },
    "dependencies": {
        "@types/node": "18.11.9"
    },
    "devDependencies": {
        "tsx": "3.11.0",
        "typescript": "4.8.4"
    }
}
```

**./test/someTests.ts**

```ts
import assert from 'assert/strict';
import { describe, it } from 'node:test';

describe('tests', () => {
    it('passes', () => {
        assert.ok(true);
    });
});
```

The test script fails with

> Could not find '/home/.../repository/test/**/*Tests.ts'

Did I miss something?

**Comment by webpro** [comment]
Another option to run the same command on Windows as well is to use [globstar](https://www.npmjs.com/package/globstar):

```json
{
  "scripts": {
    "test": "globstar -- node --loader tsx --test \"test/**/*.spec.ts\""
  }
}
```

(Last release of this package seems to be 8 years ago, but still does the job.)

**Comment by Ethan-Arrowood** [comment]
@scottwillmoore I was trying out your example and in Node v20, it now breaks with: 

```
node-test-with-typescript on  main [!] via 🤖 v20.0.0 on ☁️  production 
❯ npm test  

> test
> node --loader tsx --no-warnings script/test.ts

/Users/ethanarrowood/Documents/github/vercel/vercel-azure-devops-extension/node-test-with-typescript/script/test.ts:1
import { execSync } from "node:child_process";
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1187:20)
    at Module._compile (node:internal/modules/cjs/loader:1231:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1321:10)
    at Module.load (node:internal/modules/cjs/loader:1125:32)
    at Module._load (node:internal/modules/cjs/loader:965:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:165:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:192:25)

Node.js v20.0.0
```

**Comment by scottwillmoore** [comment]
You're right! This does not happen in Node 19 with `19.0.1`, or even `19.9.0`, but it does happen in Node 20 with `20.0.0`. The error message appears to indicate that the TypeScript is not getting transpiled into CommonJS, but instead to ModuleJS. But, with no changes to the loader [`tsx`](https://github.com/esbuild-kit/tsx) and other dependencies, this must have been caused by a change in Node 20? However, I couldn't find any concrete reason in the [changelog](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V20.md#20.0.0) for Node 20.

The solution, as indicated with @Ethan-Arrowood is to just add `"type": "module"` to your `package.json`, therefore Node will accept the TypeScript that is transpiled to ModuleJS. I've updated [my example repository](https://github.com/scottwillmoore/node-test-with-typescript) to reflect these changes. To be honest, if you are using the latest version of Node, you probably should use ModuleJS.

I've also double checked, and [a custom script](https://github.com/scottwillmoore/node-test-with-typescript/blob/main/script/test.ts) is still required to run TypeScript with the Node test runner at the moment. This is due to the [current Node test runner execution model](https://nodejs.org/docs/latest-v20.x/api/test.html#test-runner-execution-model). However, there is a feature request to fix this issue: nodejs/node#46292.

**Comment by reedHam** [comment]
Hmm when I try `node --loader tsx --test` with node 20 on windows. 
I'm getting this error. 
```
node:internal/worker:211
    this[kHandle] = new WorkerImpl(url,
                    ^
This Environment was initialized without a V8::Inspector
Thrown at:
    at Worker (node:internal/worker:211:21)
    at InternalWorker (node:internal/worker:464:5)
    at HooksProxy (node:internal/modules/esm/hooks:496:20)
    at CustomizedModuleLoader (node:internal/modules/esm/loader:351:18)
    at createModuleLoader (node:internal/modules/esm/loader:421:14)
    at get esmLoader (node:internal/process/esm_loader:15:26)
    at node:internal/modules/cjs/loader:119:47
    at node:internal/util:764:15
    at Module.load (node:internal/modules/cjs/loader:1128:26)
    at Module._load (node:internal/modules/cjs/loader:965:12)
```

**Comment by MoLow** [comment]
> Hmm when I try `node --loader tsx --test` with node 20 on windows. I'm getting this error.
> 
> ```
> node:internal/worker:211
>     this[kHandle] = new WorkerImpl(url,
>                     ^
> This Environment was initialized without a V8::Inspector
> Thrown at:
>     at Worker (node:internal/worker:211:21)
>     at InternalWorker (node:internal/worker:464:5)
>     at HooksProxy (node:internal/modules/esm/hooks:496:20)
>     at CustomizedModuleLoader (node:internal/modules/esm/loader:351:18)
>     at createModuleLoader (node:internal/modules/esm/loader:421:14)
>     at get esmLoader (node:internal/process/esm_loader:15:26)
>     at node:internal/modules/cjs/loader:119:47
>     at node:internal/util:764:15
>     at Module.load (node:internal/modules/cjs/loader:1128:26)
>     at Module._load (node:internal/modules/cjs/loader:965:12)
> ```

it seems to be caused by https://github.com/nodejs/node/blob/8becacb25d3c275341a9fef1a94581de3bd60c3d/src/env-inl.h#L638-L641

@nodejs/loaders @nodejs/workers do workers require a v8 inspector in order to work?

**Comment by JakobJingleheimer** [comment]
This sounds familiar

https://github.com/nodejs/node/pull/44710#issuecomment-1336332128

I think it ended up being a red-herring.

**Comment by webpro** [comment]
A potential solution to the OP is to pre-compile TS to JS. Then you don't need a loader and with more than two unit tests it becomes a lot faster too, since you compile everything only once and no further overhead. ([example commit](https://github.com/webpro/knip/commit/1e13de331b58f49520bb779ed125565d1fa1463d) that does this using swc).

**Comment by GeoffreyBooth** [comment]
> I think it ended up being a red-herring.

Are you sure? I think it was a red herring _for that PR_ in that Node’s internal test runner doesn’t use `--test`, so the command you were using for one-off testing in that comment wasn’t working but the actual Node tests were passing. But I think maybe it really is a bug that `--test` and `--loader` can’t be used together at the moment, because of https://github.com/nodejs/node/blob/8becacb25d3c275341a9fef1a94581de3bd60c3d/src/env-inl.h#L640. cc @MoLow

**Comment by scottwillmoore** [comment]
Weird. I have Windows 11 Pro (Build 22621) and in PowerShell with Node 20 installed with [Volta](https://volta.sh/) I don't have any issues.

````ini
git clone git@github.com:scottwillmoore/node-test-with-typescript
cd ./node-test-with-typescript

volta --version # 1.1.1
node --version # v20.0.0
npm --version # 9.6.4

npm clean-install

# Indirect
npm run test # node --loader tsx --no-warnings ./scripts/test.ts

# Direct
node --loader tsx --no-warnings --test ./tests/add.test.ts
````

Could @MoLow try with this example, or make their example reproducible.

**Comment by MoLow** [comment]
>  But I think maybe it really is a bug that --test and --loader can’t be used together 

@GeoffreyBooth is v8 inspector a requirement for using workers?

**Comment by GeoffreyBooth** [comment]
> > But I think maybe it really is a bug that –test and –loader can’t be used together
> 
> @GeoffreyBooth is v8 inspector a requirement for using workers?

Well I think most users would assume that they can attach the debugger to worker threads, so when we run v8 for a worker it expects to have that feature enabled?

**Comment by Tobbe** [comment]
https://github.com/isaacs/node-glob recently got [cli capabilities](https://github.com/isaacs/node-glob/commit/94a36ce995727fcbef8318977832e75e305786ed), so this is what I do now
```
"test:node": "yarn glob './src/**/__tests__/*.test.mts' --cmd='yarn node --loader tsx --test-reporter spec --test'",
```

**Comment by scottwillmoore** [comment]
Awesome, thanks for the heads up. I've updated [my example](https://github.com/scottwillmoore/node-test-with-typescript/tree/main) to use this approach. You can use the command below if you don't want to require Yarn.

```json
"test": "glob -c \"node --loader tsx --no-warnings --test\" \"./tests/**/*.test.ts\""
```

**Comment by nohehf** [comment]
Hi, following back on this, is it possible to pass flags to tsx when using it as a node loader? I need to specify a different tsconfig for testing. Thanks!

**Comment by joshunrau** [comment]
Also another solution for those for non-windows users who don't want to install additional dependencies:

```json
"test": "find ./src -name '*.spec.ts' -exec node --loader @swc-node/register/esm --test {} \\;"
```

**Comment by jasonwilliams** [comment]
> 

I've followed this and I get `SyntaxError: Cannot use import statement outside a module`, am i doing something wrong?

**Comment by mintyPT** [comment]
> Also another solution for those for non-windows users who don't want to install additional dependencies:
> 
> ```json
> "test": "find ./src -name '*.spec.ts' -exec node --loader @swc-node/register/esm --test {} \\;"
> ```

This sent me in the right direction but I wanted to pass all the files to the command instead of one by one. So here's what worked for me:

```json
"test": "find ./test -name '*Tests.ts' | tr '\\n' ' ' | xargs node --require ts-node/register --test "}
```

**Comment by jasonwilliams** [comment]
Just seen above https://github.com/nodejs/help/issues/3902#issuecomment-1519018966 that the fix for the syntax issue is to set “type”:”module” in package.json 

Is there a workaround for those of us who can’t do that? This is a big project, and most of the files in this project are still commonjs and that can’t really change due to other reasons (upstream electron not supporting ESM right now).

is there not a way to have tsx output commonjs instead?

@scottwillmoore

**Comment by jasonwilliams** [comment]
Thanks @scottwillmoore 

In case anyone else comes here, I've managed to get it working by doing `node --require @esbuild-kit/esm-loader  --test  test/**/x.t.ts` instead. This at least does work for now. 

It seems like the underlying issue is a [Node v20 breaking change](https://github.com/TypeStrong/ts-node/issues/1997#issuecomment-1517388341)

If I run `node --require @esbuild-kit/esm-loader  --test  test/**/x.t.ts` it also fails, so the problem is narrowed down there. Maybe the esm loader is doing something which was previously dependent on environment information and now it fails.

**Comment by robertsLando** [comment]
Giving that this costed me hours of time I will share my finding.

Easy setup:

```
npm install -D esbuild-register
node --test -r esbuild-register test/*.ts
```

Using custom `runTests.ts` see [here](https://github.com/mqttjs/MQTT.js/blob/refactor-tests/test/runTests.ts):

```ts
import { readdirSync } from 'node:fs'
import { run } from 'node:test'
import process from 'node:process'
import { spec as Spec } from 'node:test/reporters'
import { basename } from 'node:path'
import { cpus } from 'node:os'

const spec = new Spec()

let exitCode = 0

// find test files, this can be replaced using `glob` package
const files = readdirSync(__dirname)
	.filter((f) => f.endsWith('.ts') && f !== basename(__filename))
	.map((f) => `${__dirname}/${f}`)

const testStream = run({
	files,
	timeout: 60 * 1000,
	concurrency: cpus().length,
})

// set the test-reporter to `spec`, you can also use `tap` or `dot` 
testStream.compose(spec).pipe(process.stdout)

// keep track of errors/failed tests
const summary: string[] = []

testStream.on('test:fail', (data) => {
	exitCode = 1
	const error = data.details.error
	summary.push(
		`✖ ${data.file} - Test "${data.name}" failed in ${Math.round(
			data.details.duration_ms,
		)}ms: ${error.message}\n${error.stack} `,
	)
})

testStream.on('test:stderr', (data) => {
	summary.push(`${data.file} - Error:\n${data.message} `)
})

testStream.once('end', () => {
	if (summary.length > 0) {
		console.error('--- ERRORS SUMMARY ---\n')
		console.error('\x1b[31m%s\x1b[0m', summary.join('\n'), '\x1b[0m\n')
		console.error('--- END OF SUMMARY ---')
	}
	// IMPORTANT: if at least one test has an error set the correct exit code so the action will fail on ci
	process.exit(exitCode)
})
```

**Comment by JustinGrote** [comment]
Quick note that @robertsLando's example does not work on NodeJS `18.15` due to the reporters not being exposed. This is what is currently used in Electron 25 and subsequently, VSCode 1.82

**Comment by robertsLando** [comment]
@JustinGrote could you explain me what specifically doesn't work with my example in nodejs 18? I have nodejs 18 and 20 in my ci and they both work

**Comment by JustinGrote** [comment]
@robertsLando  `18.15` specifically, some of the items like the compose() function on stream and test/reporters being accessible publicly were not available yet, mostly added in `18.17`

**Comment by umstek** [comment]
I really tried but without any luck. Finally resorted to `run-tests.ts`
```typescript
import { run } from "node:test";
import { spec } from "node:test/reporters";
import process from "node:process";

import { glob } from "glob";

const tsTests = await glob("src/**/*.test.ts");

run({ files: tsTests }).compose(new spec()).pipe(process.stdout);
```
`.swcrc`
```json
{
  "$schema": "https://json.schemastore.org/swcrc",
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": false
    }
  }
}
```

and  `package.json`
```jsonc
{
  // ...
  "type": "module",
  "scripts": {
    // ...
    "test": "SWCRC=true node --loader @swc-node/register/esm run-tests.ts"
    // ...
  }
  // ...
}
```

Was it worth the effort? IDK.

**Comment by SanderPs** [comment]
After a lot of trial and error I got this to work on Windows:

```
  "scripts": {
    "test": "glob -c \"node --test --require ts-node/register\" \"./tests/**/*.ts\""
  },
```

**Comment by Dias1c** [comment]
For run all test files in you project by "node" on typescript, run command bellow:
```sh
# !important: This works only on linux
find -name "*test.ts" | xargs node --loader tsx --test
```

If it works, add it to scripts in `package.json`

**Comment by JustinGrote** [comment]
@Dias1c pretty sure that's why people are using glob, it's cross platform :)

**Comment by maestart** [comment]
For new Node versions. The `--loader` flag was deprecated in Node v20.6.0, use `--import` flag. 

This worked for me: `node --import tsx --test ./tests/*test.ts`

**Comment by jensbodal** [comment]
You can just use 

```
"test": "tsx --test src/*.test.ts"
```

The real misunderstanding seems to be how default globbing works with nodejs. If you're using `zsh` for example and run 

```
tsx --test src/**/*.test.ts
```

It will run fine. But if you use bash the shell it will fail. I assume the default globbing in nodejs uses the same backwards compatible syntax in linux where `*` and `**` are equivalent. 

So to run all tests under `src` and all subfolders is not possible without either

1. Using a glob package that supports `**`
    ```
    "test": "glob src/**/*.test.ts -c 'tsx --test'"
    ```
2. Using something like `find` to find all the test files
    ```
    "test": "tsx --test $(find src -type f -name '*.test.ts')"
    ```
3. Manually including all possible nesting levels for folders 
    ```
    "test": "tsx --test src/*.test.ts src/*/*.test.ts src/*/*/*/*.test.ts"
    ```

**Comment by varunkamra** [comment]
Hey guys, is anyone having issues with environment variables not being defined with node test runner? I am running my test with this command
'node --experimental-test-coverage --no-warnings --test -r ts-node/register tests/api/base.test.ts`
When I run this and try to console log process.env it's not defined. However, when I try to run a normal .ts script with command `node -r ts-node/register testenv.ts` and log process.env, it is able to log all environment variables.

**Comment by zauni** [comment]
> Awesome, thanks for the heads up. I've updated [my example](https://github.com/scottwillmoore/node-test-with-typescript/tree/main) to use this approach. You can use the command below if you don't want to require Yarn.
> 
> ```json
> "test": "glob -c \"node --loader tsx --no-warnings --test\" \"./tests/**/*.test.ts\""
> ```

Currently you can even shorten it to

```json
"test": "glob -c \"tsx --test\" \"./src/**/*.test.ts\""
```

**Comment by BoscoDomingo** [comment]
I just made a succinct guide to tie all of the teachings mentioned here together, including .env file and correct TS language server support in test files, all compatible with monorepos.

Feel free to check [my blog post](https://dev.to/boscodomingo/the-ultimate-node-test-runner-typescript-env-variables-guide-1hlj) and [Gist](https://gist.github.com/BoscoDomingo/a06329ca1a2f36ccb7c168d4a8de6c92) that explain it all (both free, I make no money from either, it was just easier to share the info there for whomever may be interested :) )

TL;DR:
```json
"scripts": {
    //...
    "test": "dotenv -e .env -e env.test -- glob -c \"tsx --test --test-reporter spec \" \"./test/**/*.test.ts\"",
    "build": "tsc -p tsconfig.build.json",
```

**Comment by gregg-cbs** [comment]
node: "we now support typescript and built in tests"
also node: "but its not possible to run tests that are typescript files"

**Comment by nathanjhood** [comment]
> node: "we now support typescript and built in tests" also node: "but its not possible to run tests that are typescript files"

It absolutely is possible (unless I've misunderstood you):

```ts
#!/usr/bin/env -S tsx

import * as test  from 'node:test';

import setupTests from './setupTests'

type NodeTestRunnerParameters = Required<Parameters<typeof test.run>>;
type NodeTestRunnerOptions = NodeTestRunnerParameters[0];
type NodeTestRunnerReturnType = ReturnType<typeof test.run>;

const run = (proc: NodeJS.Process, options?: NodeTestRunnerOptions) => {
  proc.on('unhandledRejection', (error) => {
    throw error;
  });
  // ... 
  return test.run(options)
}

if (require.main === module) {
  const testStream = runTests(global.process, setupTests);
}
```

```ts
// setupTests.ts
import test = require('node:test');
import reporters = require('node:test/reporters');
import path = require('node:path');
import os = require('node:os');

type NodeTestRunnerParameters = Required<Parameters<typeof test.run>>;
type NodeTestRunnerOptions = NodeTestRunnerParameters[0];

const options: NodeTestRunnerOptions = {
  // use either 'files' or 'testNamePatterns' - not both...
  files: [
    path.resolve(path.join(__dirname, '/path/to/someModuleA.test.ts')),
    path.resolve(path.join(__dirname, '/path/to/someModuleB.test.ts')),
  ],
  // testNamePatterns: [
  //   "**/*.test.?(c|m)js",
  //   "**/*-test.?(c|m)js",
  //   "**/*_test.?(c|m)js",
  //   "**/test-*.?(c|m)js",
  //   "**/test.?(c|m)js",
  //   "**/test/**/*.?(c|m)js"
  // ],
  concurrency: os.availableParallelism() - 1,
  setup(testsStream) {
    // Log test failures to console
    testsStream.on('test:fail', (testFail) => {
      console.error(testFail);
      process.exitCode = 1; // must be != 0, to avoid false positives in CI pipelines
    });
    // coverage reporter
    const isTTY = process.stdout.isTTY;
    const reporter = isTTY ? reporters.spec : reporters.tap;
    testsStream.compose(reporter).pipe(process.stdout);
  },
  // ... 
} satisfies NodeTestRunnerOptions;
```

```sh
chmod +x ./runTests.ts

# then...

./runTests.ts
```

*or*

```sh
yarn tsx ./runTests.ts
```

Apart from using [tsx](https://tsx.is/), there are [other ways](https://nodejs.org/en/learn/typescript). But as it stands, tsx is the way to go for testing Typescript-based NodeJs projects, IMO.

**Comment by marco-ippolito** [comment]
I completely missed this conversation. This feature has been supported for a while.

**Comment by darcyrush** [comment]
So I tried using the following on Node 22
```
node --require ts-node/register --test 'test/**/*.test.ts'",
``` 
And while it successfully runs most of the time, its maxes out every core on my 32 core processor and total execution time is relatively slow. This means TTD is not an option without throttling my workstation. The other times at least one test fails with a segmentation fault.

> It absolutely is possible 

So I implemented the code above thinking it might improve things, but I have the exact same behavior.

Instead of using the code above or using ts-node as a loader, if I use the `tape` package with ts-node directly, I can see that only five cores are used, at about 60% each and test execution is 33% faster. I still get the expected console output, but just with some extra empty tape output after...

I don't have the time or expertise to dive into what exactly the issue is (if it's `ts-node` specifically or something else) and I understand these features are in their early stages - I just wanted to note my experiences so far. Unfortunately, I will need to stick to using a third-party library as a test runner for now.
